File: backend/entity/Scholarship.go
package entity

import "gorm.io/gorm"

type Scholarship struct {
	gorm.Model
	ScholarshipName string `gorm:"not null" json:"scholarship_name" valid:"required~Scholarship name is required,stringlength(2|255)~Scholarship name must be 2-255 characters"`
	Description     string `gorm:"not null" json:"description" valid:"required~Description is required,stringlength(5|1000)~Description must be 5-1000 characters"`
	OpenDate        string `gorm:"not null" json:"open_date" valid:"required~Open date is required"`
	CloseDate       string `gorm:"not null" json:"close_date" valid:"required~Close date is required"`

	StatusscholarshipID uint              `json:"statusscholarship_id" valid:"required~Status scholarship is required"`
	Statusscholarship   Statusscholarship `gorm:"foreignKey:StatusscholarshipID" json:"statusscholarship"`
 
	TypescholarshipID uint            `json:"typescholarship_id" valid:"required~Type scholarship is required"`
	Typescholarship   Typescholarship `gorm:"foreignKey:TypescholarshipID" json:"typescholarship" valid:"-"`

	SemasterID uint     `json:"semaster_id" valid:"required~Semaster is required"`
	Semaster   Semaster `gorm:"foreignKey:SemasterID" json:"semaster" valid:"-"`

	SponsorID uint `json:"sponsor_id" valid:"required~Sponsor is required"`
	Sponsor Sponsor `gorm:"foreignKey:SponsorID" valid:"-"`
	
	ApprovalRequirements    []ApprovalRequirement    `gorm:"foreignKey:ScholarshipID" json:"approval_requirements" valid:"-"`
	ApplicationScholarships []ApplicationScholarship `gorm:"foreignKey:ScholarshipID" json:"application_scholarships" valid:"-"`

	InterviewRounds         []InterviewRound         `gorm:"foreignKey:ScholarshipID" json:"interview_rounds" valid:"-"`
}

File: frontend/src/interfaces/scholarship.ts
import type { SemasterResponse } from "./semaster";
import type { SponsorResponse } from "./sponsor"

export interface ScholarshipStatusResponse {
  ID: number
  status_name: string
}

export interface ScholarshipStatusCreate {
  status_name: string
}

export interface ScholarshipStatusUpdate {
  ID: number
  status_name?: string
}

export interface ScholarshipTypeResponse {
  ID: number
  type_name: string
}

export interface ScholarshipTypeCreate {
  type_name: string
}

export interface ScholarshipTypeUpdate {
  ID: number
  type_name?: string
}

export interface RequirementResponse {
  ID: number;
  name: string;
}

export interface ApprovalRequirementResponse {
  ID: number
  scholarship_id: number
  requirement_id: number
  requirement: RequirementResponse
}

export interface ScholarshipResponse {
  ID: number
  scholarship_name: string
  description: string
  open_date: string
  close_date: string
  
  statusscholarship_id: number;
  statusscholarship: ScholarshipStatusResponse

  typescholarship_id: number;
  typescholarship: ScholarshipTypeResponse

  sponsor_id: number
  sponsor: SponsorResponse

  semaster_id: number
  semaster: SemasterResponse
  

  approval_requirements: ApprovalRequirementResponse[]
}

export interface ScholarshipCreate {
  scholarship_name: string;
  description: string;
  open_date: string;
  close_date: string;

  statusscholarship_id: number;
  typescholarship_id: number;

  sponsor_id: number

  semaster_id: number
  
}

export interface ScholarshipUpdate {
  ID: number;
  scholarship_name?: string;
  description?: string;
  open_date?: string;
  close_date?: string;

  statusscholarship_id?: number;
  typescholarship_id?: number;

  sponsor_id?: number

  semaster_id?: number
}